File: Data Structures & Algorithms/string-encode-and-decode/submission-4.py
class Solution:
    def encode(self, strs: List[str]) -> str:
        encoded = ""

        for string in strs:
            encoded += str(len(string)) + "#" + string
        
        return encoded
 
    def decode(self, s: str) -> List[str]:
        ans = []
        i = 0
        while i < len(s):
            num = 0
            while s[i] != "#":
                num = num * 10 + int(s[i])
                i += 1
                    
            ans.append(s[i + 1 : i + num + 1])
            i = i + num + 1

        return ans

File: Data Structures & Algorithms/anagram-groups/submission-4.cpp
#include <vector>
#include <unordered_map>
#include <map>
#include <string>

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        std::vector<std::vector<std::string>> ans;
        std::map<std::vector<int>, std::vector<std::string>> anagrams;

        for (auto& str : strs) {
            std::vector<int> freq(26, 0);
            for (auto chr : str) {
                freq[chr - 'a'] += 1;
            }
            anagrams[freq].push_back(str);
        }

        for (auto& [k, v] : anagrams) {
            ans.push_back(v);
        }

        return ans;
    }
};
